encoder_decoder.go

```diff
@@ -8,16 +8,18 @@ type encoder interface {
 	encode(pe packetEncoder) error
 }
 
+type encodeFunc func(pe packetEncoder) error
+
 // Encode takes an Encoder and turns it into bytes.
-func encode(in encoder) ([]byte, error) {
-	if in == nil {
+func encode(ef encodeFunc) ([]byte, error) {
```

Comment:
The function can be decorated with another function that is how I take `request.encoder` and decorated it with `encodeRequest` function to add header fields during encoding.